test/functional/cmdLineTests/criu/src/org/openj9/criu/TimeChangeTest.java

```diff
@@ -21,29 +21,162 @@
-	// The elapsed time is expected less than 2 second.
-	private static long MAX_ELAPSED_TIME = 2000 * 1000 * 1000;
-	private static Path imagePath = Paths.get("cpData");
+	// maximum tardiness - 4 second
+	private final static long MAX_TARDINESS_MS = 4000;
+	private final static long MAX_TARDINESS_NS = MAX_TARDINESS_MS * 1000 * 1000;
+	// delay 5000 millisecond
+	private final static long MILLIS_DELAY = 5000;
+	private final long startNanoTime = System.nanoTime();
+	private final long currentTimeMillis = System.currentTimeMillis();
+	private final static Path imagePath = Paths.get("cpData");
 
-	public static void main(String args[]) {
-		new TimeChangeTest().testSystemNanoTime();
+	public static void main(String args[]) throws InterruptedException {
+		if (args.length == 0) {
+			throw new RuntimeException("Test name required");
+		} else {
+			String testName = args[0];
+			TimeChangeTest tct = new TimeChangeTest();
+			if ("testSystemNanoTime".equalsIgnoreCase(testName)) {
+				tct.testSystemNanoTime();
+			} else {
+				tct.test(testName);
+			}
+		}
+	}
+
+	private void test(String testName) throws InterruptedException {
+		System.out.println("Start test name: " + testName);
+		showThreadCurrentTime("Before starting " + testName);
+		Timer timer = new Timer();
+		switch (testName) {
+		case "testMillisDelay":
+			testMillisDelay(timer);
+			break;
+		case "testDateScheduledBeforeCheckpointDone":
+			testDateScheduledBeforeCheckpointDone(timer);
+			break;
+		case "testDateScheduledAfterCheckpointDone":
+			testDateScheduledAfterCheckpointDone(timer);
+			break;
+		default:
+			throw new RuntimeException("Unrecognized test name: " + testName);
+		}
+		CRIUTestUtils.checkPointJVM(imagePath, false);
+		// maximum test running time is 12s
+		Thread.sleep(12000);
+		showThreadCurrentTime("End " + testName);
+		timer.cancel();
 	}
 
 	public void testSystemNanoTime() {
 		final long beforeCheckpoint = System.nanoTime();
 		System.out.println("System.nanoTime() before CRIU checkpoint: " + beforeCheckpoint);
-		CRIUTestUtils.checkPointJVM(imagePath);
+		CRIUTestUtils.checkPointJVM(imagePath, false);
 		final long afterRestore = System.nanoTime();
 		final long elapsedTime = afterRestore - beforeCheckpoint;
-		if (elapsedTime < MAX_ELAPSED_TIME) {
-			System.out.println("PASSED: System.nanoTime() after CRIU restore: " + afterRestore + ", the elapse time is: " + elapsedTime);
+		if (elapsedTime < MAX_TARDINESS_NS) {
+			System.out.println("PASSED: System.nanoTime() after CRIU restore: " + afterRestore
+					+ ", the elapse time is: " + elapsedTime + " ns");
 		} else {
-			System.out.println("FAILED: System.nanoTime() after CRIU restore: " + afterRestore + ", the elapse time is: " + elapsedTime);
+			System.out.println("FAILED: System.nanoTime() after CRIU restore: " + afterRestore
+					+ ", the elapse time is: " + elapsedTime + " ns, w/ MAX_TARDINESS_NS : " + MAX_TARDINESS_NS);
 		}
 	}
+
+	private final TimerTask taskMillisDelay = new TimerTask() {
```

Comment:
will there be a test added for the case where a periodic task is scheduled with with a period much less than the time the JVM was shutdown due to checkpoint. Ie period is 100ms, JVM downtime is 2secods ?

Reply:
Is `taskDateScheduledBeforeCheckpointDone` close to the test scenario requested?

Reply:
Added `taskMillisDelayBeforeCheckpointDone` & `testMillisDelayAfterCheckpointDone`. These are non-repeating scheduled tasks which I hope are sufficient if the first `run` is launched as expected and following runs just follow it. Ideally a restore image can run existing sanity test suites w/o re-inventing the wheels but we are not there yet.